Modules/Core/Common/include/itkMetaDataObject.h

```diff
@@ -134,7 +134,7 @@ class ITK_TEMPLATE_EXPORT MetaDataObject : public MetaDataObjectBase
   friend bool
   operator==(const Self & lhs, const Self & rhs)
   {
-    return lhs.m_MetaDataObjectValue == rhs.m_MetaDataObjectValue;
+    return EqualMetaDataObjects(&lhs.m_MetaDataObjectValue, &rhs.m_MetaDataObjectValue);
```

Comment:
I would consider passing these in by reference rather than as `const` pointers.  That mimics typical `operator==` convention.  Also, it discourages `nullptr` as an argument to `EqualMetaDataObjects`.

Reply:
@Leengit  In general, I also prefer pass by reference rather than by pointer. But `EqualMetaDataObjects` is just an internal private helper function overload set, so it does not matter so much.  And in this case it is actually _essential_ that the parameters are pointer! The idea is that by providing an overload for void pointers, the compiler must do overload resolution. The overload for void pointers is less specific than the one for `const TMetaDataObject *`, so the compiler will always select the one for `const TMetaDataObject *`... except when the one for one for `const TMetaDataObject *` is SFINAE-ed away!

So only when SFINAE has _removed_ the `const TMetaDataObject *` specific overload, the compiler will select the void pointer overload to be called. OK?

SFINAE removes the `const TMetaDataObject *` specific overload when the expression `decltype(*lhs == *rh)` is invalid ("substitution failure") for that specific `TMetaDataObject` type.